### PR TITLE
Fix ASGIGetter keys to fetch from actual carrier headers

### LINKED ISSUE #1487: ASGI Instrumentation ValueError: too many values to unpack (expected 2) during JaegerPropagator
**Describe your environment**
- python:3.11-slim docker image
- OTEL
```bash
opentelemetry-api                           1.14.0
opentelemetry-exporter-jaeger-thrift        1.14.0
opentelemetry-instrumentation               0.35b0
opentelemetry-instrumentation-asgi          0.35b0
opentelemetry-instrumentation-fastapi       0.35b0
opentelemetry-propagator-jaeger             1.14.0
opentelemetry-sdk                           1.14.0
opentelemetry-semantic-conventions          0.35b0
opentelemetry-util-http                     0.35b0
```

**Steps to reproduce**
1. Setup an upstream application that previously create the jaeger header during the original request(example: Traefik)
2. Set up JaegerPropagator and Instrumentation on FastAPI

example Traefik config
```
  traefik:
    image: traefik:2.6
    restart: always
    command:
      --api.dashboard=true
      --providers.docker=true
      --providers.docker.exposedbydefault=false
      --entrypoints.web.address=:80
      --entrypoints.web.http.redirections.entryPoint.to=websecure
      --entrypoints.web.http.redirections.entryPoint.scheme=https
      --entrypoints.websecure.address=:443
      --accesslog=true
      --tracing.jaeger=true
      --tracing.jaeger.gen128Bit=true
      --tracing.jaeger.samplingServerURL=http://jaeger:5778/sampling
      --tracing.jaeger.samplingType=const
      --tracing.jaeger.samplingParam=1.0
      --tracing.jaeger.localAgentHostPort=jaeger:6831
    ports:
      - 80:80
      - 443:443
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    labels:
      - traefik.enable=true
      - traefik.http.routers.dashboard.rule=Host(`traefik.docker.localhost`) && (PathPrefix(`/api/`) || PathPrefix(`/dashboard/`))
      - traefik.http.routers.dashboard.tls=true
      - traefik.http.routers.dashboard.service=api@internal
      
  svc:
    build:
      context: .
    command: tail -f /dev/null
    restart: always
    volumes:
      - .:/usr/src
    environment:
      - SQLALCHEMY_DEBUG_ECHO=true
      - OTEL_RESOURCE_ATTRIBUTES=service.name=svc
      - OTEL_EXPORTER_JAEGER_AGENT_HOST=jaeger
      - OTEL_EXPORTER_JAEGER_AGENT_PORT=6831
      - OTEL_TRACES_SAMPLER=parentbased_always_on
      # - OTEL_TRACES_SAMPLER=parentbased_traceidratio
      # - OTEL_TRACES_SAMPLER_ARG=1.0
      - OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST=x-forwarded-for,x-real-ip
    expose:
      - 80
    labels:
      - traefik.enable=true
      - traefik.http.routers.svc.rule=Host(`api.docker.localhost`) && PathPrefix(`/`)
      - traefik.http.routers.svc.tls=true
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost/health"]
      interval: 10s
      timeout: 10s
      retries: 3
      start_period: 10s

```

```python
import fastapi
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry import trace
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.propagate import set_global_textmap
from opentelemetry.propagators.jaeger import JaegerPropagator
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

set_global_textmap(JaegerPropagator())
provider = TracerProvider()
span_processor = BatchSpanProcessor(JaegerExporter())
provider.add_span_processor(span_processor)
trace.set_tracer_provider(provider)

app = fastapi.FastAPI()


@app.get("/health")
async def health():
    return {"message": "hello world"}


FastAPIInstrumentor.instrument_app(app)
```

**What is the expected behavior?**
If the tracing already propagates with the Jaeger header(example from Traefik as the reverse proxy and with tracing enabled), we should still have the trace propagated successfully in FastAPI Instrumentation.

**What is the actual behavior?**
now even with default swagger pages, it will have exceptions on opening. ofc, I have reverse proxy both API and Documents.
- any request coming with the Jaeger header will have issues as below
- any request that does not have a previous jaeger header will still be fine (because the asgi getter will have no issue with processing empty headers)

```python
INFO:     172.20.0.7:55510 - "GET /docs HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/fastapi/applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 531, in __call__
    span, token = _start_internal_or_server_span(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/instrumentation/utils.py", line 111, in _start_internal_or_server_span
    ctx = extract(context_carrier, getter=context_getter)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/propagate/__init__.py", line 102, in extract
    return get_global_textmap().extract(carrier, context, getter=getter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/propagators/jaeger/__init__.py", line 54, in extract
    context = self._extract_baggage(getter, carrier, context)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/propagators/jaeger/__init__.py", line 119, in _extract_baggage
    for key in getter.keys(carrier)
               ^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 263, in keys
    return [_key.decode("utf8") for (_key, _value) in carrier]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssp-py-sdk/.local/lib/python3.11/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 263, in <listcomp>
    return [_key.decode("utf8") for (_key, _value) in carrier]
                                    ^^^^^^^^^^^^^^
```

**Additional context**
I believe the issue from this specific line https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L263, and I can confirm the older version 0.34b0 with this impl https://github.com/open-telemetry/opentelemetry-python-contrib/blob/v0.34b0/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L217 have no issue.

### PR DESCRIPTION
**Pull request by hasier**
# Description

Working in ASGI with any propagator using `getter.keys()`, e.g. the [GCP propagator](https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/blob/main/opentelemetry-propagator-gcp/) when no header is present in a request, it breaks with the below message:
```python
Traceback (most recent call last):
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/fastapi/applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 531, in __call__
    span, token = _start_internal_or_server_span(
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/instrumentation/utils.py", line 111, in _start_internal_or_server_span
    ctx = extract(context_carrier, getter=context_getter)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/propagate/__init__.py", line 102, in extract
    return get_global_textmap().extract(carrier, context, getter=getter)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/propagators/composite.py", line 52, in extract
    context = propagator.extract(carrier, context, getter=getter)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/propagators/cloud_trace_propagator/__init__.py", line 94, in extract
    header = self._get_header_value(getter, carrier)
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/propagators/cloud_trace_propagator/__init__.py", line 78, in _get_header_value
    for key in getter.keys(carrier):
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 263, in keys
    return [_key.decode("utf8") for (_key, _value) in carrier]
  File "/root/.cache/pypoetry/virtualenvs/api-9TtSrW0h-py3.10/lib/python3.10/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 263, in <listcomp>
    return [_key.decode("utf8") for (_key, _value) in carrier]
ValueError: too many values to unpack (expected 2)
```

In my case, this seems to be triggered by the fact that the GCP propagator [uses `getter.keys()` as a fallback](https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/blob/main/opentelemetry-propagator-gcp/src/opentelemetry/propagators/cloud_trace_propagator/__init__.py#L78). This [`keys()` function in the `ASGIGetter`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/be4ceec08cb1b3097d7aa25c4efc358ddd050bdb/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L262-L263) seems to be taking the header values from an incorrect place, and therefore provoking the crash. I found it with the GCP example, but it's probably [affecting others](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435#issuecomment-1316001798), as the issue lays with the definition of `ASGIGetter.keys()` here.

Fixes #1487 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Using any propagator that makes use of `getter.keys()` on an ASGI context makes the request fail. It's easy to reproduce with the GCP `CloudTraceFormatPropagator` when the request does not have a `x-cloud-trace-context` header.
- The added test changes make tests fail if the fix is not included.

**NOTE**: I've [re-created](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435/files#diff-cae4643dfa4b180a8069a7c31a4ead910a234dd37059d838e5ae8a6479023392R28-R79) `opentelemetry.test.mock_textmap.MockTextMapPropagator` with a tiny modification to use `keys()` from the `getter`. I understand it might be interesting to add these changes to [the original testing repo](https://github.com/open-telemetry/opentelemetry-python/blob/5574f7ea6b63c0fd7f5b32be4b0ff030f29b9efe/tests/opentelemetry-test-utils/src/opentelemetry/test/mock_textmap.py#L57-L104) as well, as it ensures all keys that `.get()` returns are in `.keys()` as well, but thought it'd be interesting to have this fix out soon and then potentially port it there if it seemed useful/appropriate to y'all.

# Does This PR Require a Core Repo Change?

- ~Yes. - Link to PR:~
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- ~Documentation has been updated~

### CODE REVIEW

**Comment by bluesliverx** [comment]
We just hit this same problem, not in GCP, and it's forcing us to downgrade to 0.34b.

**Comment by ocelotl** [comment]
> gcp/src/opentelemetry/propagators/cloud_trace_propagator/__init__.py#L78). This `all()` call in the `ASGIGetter` seems to be taking the header values from an incorrect place, and therefore provoking the crash.

Sorry, which `all()` call?

**Comment by hasier** [comment]
> > gcp/src/opentelemetry/propagators/cloud_trace_propagator/**init**.py#L78). This `all()` call in the `ASGIGetter` seems to be taking the header values from an incorrect place, and therefore provoking the crash.
> 
> Sorry, which `all()` call?

Ah sorry @ocelotl, I meant the `keys()` function, just updated the description 🙏

**Comment by hasier** [comment]
@ocelotl is this PR something you can help with reviewing? Or maybe @rogersd @srikanthccv, as I've seen you working in the ASGI area recently? Sorry for the wide ping, it just feels to me like an easy fix for a bug that can become quite common and important.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:263]: Just curious, if `"headers"` is not in `carrier.keys`, shouldn't this method return `[_key.decode("utf8") for (_key, _value) in carrier]`. The way this is being modified makes this method return an empty list if `"headers"` is not in `carrier.keys`.

Comment (hasier) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:263]: Considering the `.get()` method above returns header-specific fields and its description mentions `"header name in scope"`, I assumed here we'd only want to return headers rather than the whole carrier, which in this case is the [ASGI scope](https://asgi.readthedocs.io/en/stable/specs/www.html#http-connection-scope). I'm not too sure fetching all values from the scope is correct though, there's a bunch of ASGI-specific info there and I wouldn't expect to see (nor inject) any extra info from Opentelemetry.

Now that you mention it though, I guess some people may be relying on this behaviour now, so maybe we should join both lists? Or would we treat this as a breaking change instead?

Comment (torqataNate) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:263]: I would argue that no one was depending on this because before it was trying to iterate over a dict without dict.items() which was the initial cause of the ValueError: too many values to unpack (expected 2). So I dont know that we need to combine both lists.  The above would also fail when a header isn't a byte type so it is still broken.   I had a request i was going to submit but im not a contributor.  how would you like me to submit the proposed change?

Comment (hasier) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:263]: As far as this processor is concerned, it has been working until [this change](https://github.com/open-telemetry/opentelemetry-python-contrib/commit/9d6ba63e06db78d187008f5aa8fe7fe216b270d2#diff-7eb6c4565bcdedf2510f8932a48d5781f49bdf50b38a56d7d115a76d2100e186R263). The previous behaviour was not correct either, but at least it would not crash. The changes in this PR directly address the main issue, which is the crash due to iterating the wrong value, and fetching the correct value the same way we do in `.get()`. I would argue that any more changes in the behaviour than that are out of scope and probably require a separate discussion and investigation. In any case, something for the maintainers to decide CC @ocelotl @srikanthccv

**Comment by hasier** [comment]
@ocelotl @rogersd @srikanthccv @shalevr sorry for the wide ping again, just wanted to give another nudge as it's been a few weeks since I first opened the PR. A review with improvements would be appreciated in order to get this closer to approval. Thanks.

**Comment by srikanthccv** [comment]
I will take a look at it.

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_custom_headers.py
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_custom_headers.py:165]: I couldn't follow what the tests were targeting. I would test that `ASGIGetter` instance doesn't throw an exception for scope dict, which is currently failing, and additionally assert that keys return the Headers keys.

Comment (hasier) [reply instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_custom_headers.py]: The idea was to ensure that the values returned by `get()` and `keys()` were the same and they were consistent across different tests, but it was a bit of a hack indeed with that custom propagator. I added simpler tests in `test_getter.py` instead that indeed fail before the fix and pass now, hope they make sense https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435/commits/c219bb0fd9884f7a7720a1192f480cef42379812

**Review by srikanthccv — APPROVED** [review]

**Comment by muattiyah** [comment]
Assuming the current github workflow checks pass and the PR is merged, when do the maintainers expect a release that has this included to be released?

**Comment by srikanthccv** [comment]
A periodic monthly release is happening this week (or early next week).

**Comment by Lawouach** [comment]
Hello @srikanthccv, unless I'm mistaken this has not been released yet right? Any chance this gets bumped as it's a bit of a challenge without this fix out. Thanks :)

**Comment by hasier** [comment]
> Hello @srikanthccv, unless I'm mistaken this has not been released yet right? Any chance this gets bumped as it's a bit of a challenge without this fix out. Thanks :)

afaik it's not been released indeed, last published version seems to be [v0.36b0](https://github.com/open-telemetry/opentelemetry-python-contrib/releases/tag/v0.36b0), which still contains the incorrect behaviour. In case it helps, while no new version is out, in our repos we have temporarily addressed it by overriding the default `asgi` getter.
```python
from opentelemetry.instrumentation import asgi

class ASGIGetter(asgi.ASGIGetter):
    """
    Override default getter implementation while `.keys()` fix is not out
    https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435
    """
    def keys(self, carrier: dict[str, typing.Any]) -> list[str]:
        headers = carrier.get("headers") or []
        return [_key.decode("utf8") for (_key, _value) in headers]

def setup_tracing() -> None:
    # Override default asgi_getter while `.keys()` fix is not out
    # https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435
    asgi.asgi_getter = ASGIGetter()
    ...
```

**Comment by Lawouach** [comment]
Thanks @hasier

A sound proposal that I'll have to review. I'm not using the ASGI extension directly but through FastAPI. Cheers.

**Comment by theobouwman** [comment]
Is this fix already merged @hasier ?

**Comment by hasier** [comment]
> Is this fix already merged @hasier ?

@theobouwman the fix was [merged on January 12](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1435#event-8215475429) and included as part of [Version 1.16.0/0.37b0](https://github.com/open-telemetry/opentelemetry-python-contrib/releases/tag/v1.16.0)

**Pull request merged**
